examples/pytorch/pyg/main.py

```diff
@@ -0,0 +1,188 @@
+        edge_attr_cls (EdgeAttr, optional): A user-defined
+            :class:`EdgeAttr` class to customize the required attributes and
+            their ordering to uniquely identify edges. (default: :obj:`None`)
+    """
+
+    def __init__(self, ge, node_path):
+        """Initialize DeepGNN graph store."""
+        super().__init__()
+        self.ge = ge
+        self.node_ids = np.loadtxt(node_path, dtype=np.int64)
+
+    def _put_edge_index(self, edge_index, edge_attr) -> bool:
+        """To be implemented by :class:`GraphStore` subclasses."""
+        return False
+
+    def _remove_edge_index(self, edge_attr) -> bool:
+        """To be implemented by :class:`GraphStore` subclasses."""
+        return False
+
+    def _get_edge_index(self, edge_attr):
+        """To be implemented by :class:`GraphStore` subclasses."""
+        edge_type = int(edge_attr.edge_type[1])
+        edge = self.ge.neighbors(self.node_ids, edge_type)
+        srcs = []
+        for i, e in enumerate(edge[-1]):
+            srcs.extend([self.node_ids[i]] * e)
+        edge_index = (
+            torch.Tensor(srcs).long(),
+            torch.Tensor(edge[0]).long(),
+        )
+        return edge_index
+
+    def get_all_edge_attrs(self):
+        """Obtain all edge attributes stored in the :class:`GraphStore`."""
+        output = []
+        ta = EdgeAttr(("0", "0", "0"), "coo", size=[self.ge.node_count(0), 2708])
+        output.append(ta)
+
+        return output
+
+
+def train():
+    """Train the model."""
+    model.train()
+
+    total_loss = total_examples = 0
+    for data in tqdm.tqdm(loader):
+        data = data.to(device)
+        data = data[("0", "0", "0")]
+        optimizer.zero_grad()
+
+        h = model(data.x, data.edge_index)
+
+        h_src = h[data.edge_label_index[0]]
+        h_dst = h[data.edge_label_index[1]]
+        link_pred = (h_src * h_dst).sum(dim=-1)  # Inner product.
+
+        loss = F.binary_cross_entropy_with_logits(link_pred, data.edge_label)
+        loss.backward()
+        optimizer.step()
+
+        total_loss += float(loss) * link_pred.numel()
+        total_examples += link_pred.numel()
+
+    return total_loss / total_examples
+
+
+def test():
+    for data in tqdm.tqdm(loader):
+        data = data[("0", "0", "0")]
+        pred = model(data.x, data.edge_index)
+        pred = pred.argmax(1).detach().numpy()
+        test_f1 = f1_score(data.y, pred, average="micro")
+        return test_f1
+
+
+if __name__ == "__main__":
+    ge = Cora()
+    loader = LinkNeighborLoader(
+        (
+            DeepGNNFeatureStore(ge),
+            DeepGNNGraphStore(ge, f"{ge.data_dir()}/train.nodes"),
+        ),
+        batch_size=140,
+        shuffle=True,
+        neg_sampling_ratio=1.0,
+        num_neighbors=[5, 5],
+        num_workers=1,
+        persistent_workers=True,
+        edge_label_index=(
+            ("0", "0", "0"),
+            torch.Tensor(ge.sample_edges(2708, np.array(0), SamplingStrategy.Random))
```

Comment:
Please, extract a const for this magic number, 2708.

Reply:
2708 is still not extracted as const. I believe this should be number of edges of type 0 or something like that.